### PR TITLE
feat(dash-spv): storage truncation primitives for reorg support

### LINKED ISSUE #138: feat(dash-spv): storage truncation primitives
Part of [#137](https://github.com/xdustinface/rust-dashcore/issues/137). **Phase 1.**

## Scope

Add a `truncate_above(height)` method to each storage trait, backed by `SegmentCache`. Pre-requisite for every other phase.

## Tasks

- [ ] `SegmentCache::truncate_above(height)` overwrites slots back to `I::sentinel()`, drops segments above the target, recomputes `tip_height`. Must satisfy the `debug_assert!(self.items[offset] == I::sentinel())` in `Segment::insert` at `dash-spv/src/storage/segments.rs:488`, otherwise re-storing into the truncated range panics in debug.
- [ ] `BlockHeaderStorage::truncate_above(u32)` on the trait at `dash-spv/src/storage/block_headers.rs:43`.
- [ ] `PersistentBlockHeaderStorage::truncate_above` also removes all entries from `header_hash_index` (`block_headers.rs:102`) with `height > target`. Cheap option, `retain(|_, h| *h <= target)`. Faster option, maintain a reverse `HashMap<u32, BlockHash>` and drop in O(truncated_count).
- [ ] `FilterHeaderStorage::truncate_above` on the trait at `dash-spv/src/storage/filter_headers.rs:11`.
- [ ] `FilterStorage::truncate_above` on the trait at `dash-spv/src/storage/filters.rs:12`.
- [ ] `BlockStorage::truncate_above` on the trait at `dash-spv/src/storage/blocks.rs:15`.
- [ ] Operate on the `to_persist` in-memory map first, flush on next `persist()`.

## Tests

- Truncate at exact segment boundary
- Truncate within a segment, then re-store, then re-read (verifies sentinel reset)
- Truncate above `tip_height` is a no-op
- Truncate below `start_height` returns an error (the higher-level guards in Phase 3 should prevent this, but storage defends in depth)
- Hash-index truncation: after truncating, `get_header_height_by_hash(orphaned_hash)` returns `None`
- Concurrent truncate + read via `RwLock` is safe (no torn state)

## Reference

- [`SegmentCache`](https://github.com/xdustinface/rust-dashcore/blob/dev/dash-spv/src/storage/segments.rs)
- dashd analog, [`CChain::SetTip`](https://github.com/dashpay/dash/blob/develop/src/chain.cpp) (rebuilds the active vector when the new tip is on a forked branch or backward)

### PR DESCRIPTION
**Pull request by xdustinface**
## Summary

- Add `SegmentCache::truncate_above(target_height)` primitive that drops segments above the boundary and overwrites the boundary segment's tail back to `I::sentinel()` so the truncated range can be re-stored without tripping the `Segment::insert` debug-assert.
- Add `truncate_above` to `BlockHeaderStorage`, `FilterHeaderStorage`, `FilterStorage`, `BlockStorage` traits with consistent semantics (no-op above tip, error below `start_height`).
- `PersistentBlockHeaderStorage::truncate_above` also prunes `header_hash_index` so orphaned block hashes no longer resolve.
- Orphaned segment files are queued in a `to_delete` set on `SegmentCache` and removed by the next `persist()`. No disk writes happen inside `truncate_above` itself.

Phase 1 of [#137](https://github.com/xdustinface/rust-dashcore/issues/137). Unblocks [#139](https://github.com/xdustinface/rust-dashcore/issues/139).

Closes #138

## Test plan

- [x] `cargo test -p dash-spv --lib` (450 passing, including new truncate tests)
- [x] `cargo clippy -p dash-spv --all-targets -- -D warnings`
- [x] New tests cover: within-segment truncate then re-store (sentinel-reset path), exact segment boundary, no-op above tip, error below `start_height`, hash-index pruning for `PersistentBlockHeaderStorage`, concurrent truncate + read through `RwLock`

### CODE REVIEW

File: dash-spv/src/storage/segments.rs [hunk 1]
```diff
@@ -322,9 +326,67 @@ impl<I: Persistable> SegmentCache<I> {
...
+                )));
+            }
+        }
+
+        let items_per_segment = Segment::<I>::ITEMS_PER_SEGMENT;
+        let boundary_segment_id = target_height / items_per_segment;
+        let boundary_offset = target_height % items_per_segment;
+        let max_segment_id = tip / items_per_segment;
+
+        for segment_id in (boundary_segment_id + 1)..=max_segment_id {
```
[the remark this answers is not in the document]
Comment (xdustinface) [reply dash-spv/src/storage/segments.rs:411]: Fixed in c7251bb. `get_segment_mut` now checks `to_delete` before falling through to `Segment::load`, returning a fresh empty segment when the id is queued for deletion. The next `persist` uses `atomic_write` to replace the file, so explicit pre-deletion is unnecessary. Added `test_truncate_above_then_store_into_dropped_segment_without_persist` as a regression test covering the exact reorg scenario.

File: dash-spv/src/storage/segments.rs [hunk 2]
```diff
@@ -322,9 +326,67 @@ impl<I: Persistable> SegmentCache<I> {
...
+
+        self.tip_height = Some(target_height);
+
+        Ok(())
+    }
+
     pub async fn persist(&mut self, segments_dir: impl Into<PathBuf>) {
         let segments_dir = segments_dir.into();
 
+        for id in self.to_delete.drain() {
```
[the remark this answers is not in the document]
Comment (xdustinface) [reply dash-spv/src/storage/segments.rs:426]: Fixed in 78bede8 along with the duplicate findings on lines 354 and 370. `persist` now retries failed deletions on the next call by collecting failures and re-inserting them into `to_delete`. `NotFound` is treated as success since the goal state is already met.

File: dash-spv/src/storage/segments.rs [hunk 3]
```diff
@@ -322,9 +326,67 @@ impl<I: Persistable> SegmentCache<I> {
...
+
+        for segment_id in (boundary_segment_id + 1)..=max_segment_id {
+            self.segments.remove(&segment_id);
+            self.evicted.remove(&segment_id);
+            self.to_delete.insert(segment_id);
+        }
+
+        if boundary_offset + 1 < items_per_segment {
+            let segment = self.get_segment_mut(&boundary_segment_id).await?;
+            segment.reset_above(boundary_offset);
```
[the remark this answers is not in the document]
Comment (xdustinface) [reply dash-spv/src/storage/segments.rs]: Fixed in 78bede8. See [the reply on the related thread](https://github.com/xdustinface/rust-dashcore/pull/149#discussion_r3294350529).

File: dash-spv/src/storage/segments.rs [hunk 4]
```diff
@@ -322,9 +326,67 @@ impl<I: Persistable> SegmentCache<I> {
...
+        }
+
+        if let Some(start) = self.start_height {
+            if target_height < start {
+                return Err(StorageError::WriteFailed(format!(
+                    "truncate_above({target_height}) below start_height ({start})"
+                )));
+            }
+        }
+
```
[the remark this answers is not in the document]
Comment (xdustinface) [reply dash-spv/src/storage/segments.rs:397]: Fixed in c7251bb by reordering. `truncate_above` now loads the boundary segment first so any disk I/O error is surfaced before any in-memory mutations. After the boundary load returns, only infallible in-memory operations run, so the function cannot leave the cache in a half-truncated state. This is simpler than snapshot-and-rollback and avoids the bookkeeping cost.

File: dash-spv/src/storage/blocks.rs [hunk 5]
```diff
@@ -112,6 +123,74 @@ mod tests {
...
+    }
+
+    #[tokio::test]
+    async fn test_truncate_above_tip_noop_blocks() {
+        let temp_dir = TempDir::new().unwrap();
+        let mut storage = PersistentBlockStorage::open(temp_dir.path()).await.unwrap();
+
+        storage.store_block(50, HashedBlock::dummy(50, vec![])).await.unwrap();
+        storage.truncate_above(1_000).await.unwrap();
+        assert_eq!(storage.load_block(50).await.unwrap(), Some(HashedBlock::dummy(50, vec![])));
```
[the remark this answers is not in the document]
Comment (xdustinface) [reply dash-spv/src/storage/blocks.rs:181]: Fixed in 661b6cf. Switched to `#[tokio::test(flavor = "multi_thread", worker_threads = 2)]` so the reader and writer tasks execute with genuine parallelism.

File: dash-spv/src/storage/block_headers.rs [hunk 6]
```diff
@@ -261,4 +274,50 @@ mod tests {
         assert_eq!(tip, expected_tip);
         assert_eq!(storage.get_tip_height().await, Some(4));
     }
+
```
[the remark this answers is not in the document]
Comment (xdustinface) [reply dash-spv/src/storage/block_headers.rs:291]: Fixed in 194a460. Extended `test_truncate_above_drops_index_entries_and_allows_restore` with a `persist`, `drop`, `open` cycle that re-verifies the orphaned and kept hash lookups against the freshly rebuilt index.

File: dash-spv/src/storage/segments.rs [hunk 7]
```diff
@@ -322,9 +326,67 @@ impl<I: Persistable> SegmentCache<I> {
...
+        if target_height >= tip {
+            return Ok(());
+        }
+
+        if let Some(start) = self.start_height {
+            if target_height < start {
+                return Err(StorageError::WriteFailed(format!(
+                    "truncate_above({target_height}) below start_height ({start})"
+                )));
+            }
```
[the remark this answers is not in the document]
Comment (xdustinface) [reply dash-spv/src/storage/segments.rs:395]: Fixed in 78bede8. See [the reply on the related thread](https://github.com/xdustinface/rust-dashcore/pull/149#discussion_r3294350529). Note: the gap between truncation and the next persist (process crash) is still uncovered, which matches the scope you flagged as out of scope for Phase 1.

File: dash-spv/src/storage/block_headers.rs [hunk 8]
```diff
@@ -95,6 +95,13 @@ pub trait BlockHeaderStorage: Send + Sync + 'static {
         &self,
         hash: &dashcore::BlockHash,
     ) -> StorageResult<Option<u32>>;
+
+    /// Drop all headers with `height > target_height`.
+    ///
+    /// Truncating above the current tip is a no-op; truncating below
```
[the remark this answers is not in the document]
Comment (xdustinface) [reply dash-spv/src/storage/block_headers.rs]: Pushing back: the lack of a default body is intentional. The repo's API stability policy (CLAUDE.md, "version 0.x.x ... Breaking changes may occur in minor version updates") explicitly permits this kind of breaking trait change, and a compile-time forcing function is preferable to a runtime `Err(NotSupported)` for an operation as critical as reorg-driven truncation. Silent runtime failure on a reorg would be much harder to diagnose than a compile error against a downstream implementor. `MockHeaderStorage` was updated in this PR for the same reason. Release-notes coverage is out of scope for this PR's diff.

File: dash-spv/src/storage/mod.rs [hunk 9]
```diff
@@ -319,6 +319,10 @@ impl BlockHeaderStorage for DiskStorageManager {
     ) -> StorageResult<Option<u32>> {
         self.block_headers.read().await.get_header_height_by_hash(hash).await
     }
+
```
[the remark this answers is not in the document]
Comment (xdustinface) [reply dash-spv/src/storage/mod.rs:322]: Pushing back as out of scope for this PR. [#138](https://github.com/xdustinface/rust-dashcore/issues/138) explicitly scopes Phase 1 to per-storage `truncate_above` primitives, and the parent issue [#137](https://github.com/xdustinface/rust-dashcore/issues/137) tracks coordinated reorg orchestration as Phase 3. Your own comment notes "This is a prerequisite for Phase 3 rather than Phase 1". The 0.x.x API policy permits the future Phase 3 trait addition as a separate breaking change, so locking the interface in now is not required to keep the upgrade path open. The cross-storage coordination, atomicity, and compensation-on-failure semantics belong in the Phase 3 design rather than as a forward-declared shape with no behaviour.

File: dash-spv/src/storage/segments.rs [hunk 10]
```diff
@@ -322,9 +330,76 @@ impl<I: Persistable> SegmentCache<I> {
...
+    pub async fn truncate_above(&mut self, target_height: u32) -> StorageResult<()> {
+        let tip = match self.tip_height {
+            Some(tip) => tip,
+            None => return Ok(()),
+        };
+
+        if target_height >= tip {
+            return Ok(());
+        }
+
```
[the remark this answers is not in the document]
Comment (xdustinface) [reply dash-spv/src/storage/segments.rs:389]: Fixed in [1dbce10](https://github.com/xdustinface/rust-dashcore/commit/1dbce106). Added a `StorageError::InvalidArgument(String)` variant and switched the bounds-check rejection in `truncate_above` over to it. All other `WriteFailed` sites remain genuine write-path failures.

File: dash-spv/src/storage/segments.rs [hunk 11]
```diff
@@ -184,9 +188,13 @@ impl<I: Persistable> SegmentCache<I> {
...
+        // If the segment is queued for deletion, return a fresh empty segment.
+        // The next `persist` will atomically overwrite the stale file.
+        // Otherwise, load it from disk.
         let segment = if let Some(segment) = self.evicted.remove(segment_id) {
             segment
+        } else if self.to_delete.remove(segment_id) {
+            Segment::new(*segment_id, vec![], SegmentState::Dirty)
         } else {
             Segment::load(&self.segments_dir, *segment_id).await?
         };
```
[the remark this answers is not in the document]
Comment (xdustinface) [reply dash-spv/src/storage/segments.rs:200]: Fixed in [c5225d7](https://github.com/xdustinface/rust-dashcore/commit/c5225d78). `get_item` now short-circuits when the segment id is in `to_delete` and returns `None` without entering `get_segment_mut`, so the deletion intent survives reads above the truncated tip. `get_items` is unaffected because its callers only request dense ranges below the tip. New test `test_read_above_truncated_tip_preserves_deletion` locks this in.

File: dash-spv/src/storage/segments.rs [hunk 12]
```diff
@@ -654,6 +752,144 @@ mod tests {
         );
```
[the remark this answers is not in the document]
Comment (xdustinface) [reply dash-spv/src/storage/segments.rs:788]: Fixed in [bca3b6f](https://github.com/xdustinface/rust-dashcore/commit/bca3b6f9). Added `test_truncate_within_segment_persist_reload` covering the exact path you flagged: truncate within a single segment, persist without re-storing, drop, reopen, and verify both `tip_height` and the sentinel slots. I locally reproduced the regression by removing the `changed` guard in `reset_above` and the test fails as expected.

File: dash-spv/src/storage/blocks.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (xdustinface) [reply dash-spv/src/storage/blocks.rs:181]: Restructured in [2022783](https://github.com/xdustinface/rust-dashcore/commit/20227837). You are right that the outer `tokio::sync::RwLock` serializes the reader and writer regardless of runtime flavor, so the previous "concurrent" framing was misleading. Renamed the test to `test_truncate_and_read_under_contention` and added a `tokio::sync::Barrier` so both tasks are scheduled and racing for the storage lock before either acquires it. Removing the outer lock would require making `BlockStorage::truncate_above` take `\&self` (the inner `RwLock<SegmentCache>` already provides interior mutability), which is a trait-API change out of scope for Phase 1.

File: dash-spv/src/storage/block_headers.rs (hunk 8 above)
[the remark this answers is not in the document]
Comment (xdustinface) [reply dash-spv/src/storage/block_headers.rs]: Pushing back again, same reasoning as on the [original thread](https://github.com/xdustinface/rust-dashcore/pull/149#discussion_r3294351494): the lack of a default body is intentional given the 0.x.x stability policy in CLAUDE.md, and a compile-time forcing function is preferable to a runtime `Err(NotImplemented)` that would let stale truncation behavior pass unnoticed. No new evidence has surfaced that would change that trade-off.

File: dash-spv/src/storage/segments.rs (hunk 11 above)
[the remark this answers is not in the document]
Comment (xdustinface) [reply dash-spv/src/storage/segments.rs:200]: Fixed in d90b0f81. Mirrored the `get_item` guard in the per-segment loop so a range read that spans a `to_delete` segment now returns `InvalidArgument` instead of consuming the deletion intent. Added `test_get_items_above_truncated_tip_preserves_deletion` to cover the path.

File: dash-spv/src/storage/segments.rs [hunk 13]
```diff
@@ -322,9 +339,76 @@ impl<I: Persistable> SegmentCache<I> {
...
+
+        if let Some(start) = self.start_height {
+            if target_height < start {
+                return Err(StorageError::InvalidArgument(format!(
+                    "truncate_above({target_height}) below start_height ({start})"
+                )));
+            }
+        }
+
+        let items_per_segment = Segment::<I>::ITEMS_PER_SEGMENT;
```
[the remark this answers is not in the document]
Comment (xdustinface) [reply dash-spv/src/storage/segments.rs]: Fixed in 64afa99a. `truncate_above` now uses the existing `height_to_segment_id` and `height_to_offset` helpers and references `Segment::<I>::ITEMS_PER_SEGMENT` directly at the one remaining call site.

File: dash-spv/src/storage/segments.rs [hunk 14]
```diff
@@ -322,9 +349,75 @@ impl<I: Persistable> SegmentCache<I> {
         Ok(())
```
[the remark this answers is not in the document]
Comment (xdustinface) [reply dash-spv/src/storage/segments.rs:362]: Added the durability caveat to `SegmentCache::truncate_above` and to the four `truncate_above` trait docs (`block_headers`, `blocks`, `filter_headers`, `filters`). Fixed in a7ec7fe5.

File: dash-spv/src/storage/segments.rs [hunk 15]
```diff
@@ -215,6 +223,16 @@ impl<I: Persistable> SegmentCache<I> {
         let end_segment = Self::height_to_segment_id(end - 1);
```
[the remark this answers is not in the document]
Comment (xdustinface) [reply dash-spv/src/storage/segments.rs:230]: Documented on `SegmentCache::get_items` plus the three trait-level range readers (`load_headers`, `load_filter_headers`, `load_filters`) that they return `InvalidArgument` when the range spans a segment queued for deletion, and that callers must clamp the range. Fixed in a7ec7fe5.

File: dash-spv/src/storage/block_headers.rs [hunk 16]
```diff
@@ -234,6 +241,12 @@ impl BlockHeaderStorage for PersistentBlockHeaderStorage {
     ) -> StorageResult<Option<u32>> {
```
[the remark this answers is not in the document]
Comment (xdustinface) [reply dash-spv/src/storage/block_headers.rs:250]: Adopted the simpler variant of the suggested fix: read `tip_height` before calling the inner truncate, and only run `header_hash_index.retain` when `target_height < tip` (i.e. the truncate would actually drop entries). Fixed in bf07390f.

File: dash-spv/src/storage/segments.rs [hunk 17]
```diff
@@ -654,6 +770,228 @@ mod tests {
         );
```
[the remark this answers is not in the document]
Comment (xdustinface) [reply dash-spv/src/storage/segments.rs:788]: Added two tests: `test_sequential_truncations_within_same_segment` (two cuts inside one segment, then re-store across the doubly-truncated range) and `test_sequential_truncations_across_segment_boundaries` (cuts that cross different segment boundaries, with persist + reopen to verify the on-disk file set). Fixed in ea95d45f.

File: dash-spv/src/storage/blocks.rs [hunk 18]
```diff
@@ -112,6 +127,82 @@ mod tests {
         assert!(loaded.is_none());
```
[the remark this answers is not in the document]
Comment (xdustinface) [reply dash-spv/src/storage/blocks.rs:131]: Added `test_truncate_above_persist_reopen_blocks` and `test_truncate_above_persist_reopen_filters`, mirroring the existing `PersistentBlockHeaderStorage` durability test. A boundary-segment dirty-flag regression would now surface in all three storage types. Fixed in ea95d45f.

File: dash-spv/src/storage/segments.rs [hunk 19]
```diff
@@ -261,9 +279,18 @@ impl<I: Persistable> SegmentCache<I> {
     }
 
     /// Get a single item by height. Returns `None` for sentinel (empty) slots.
-    /// Unlike `get_items()`, this does not assert dense storage — safe for sparse data.
+    /// Unlike `get_items()`, this does not assert dense storage, safe for sparse data.
     pub async fn get_item(&mut self, height: u32) -> StorageResult<Option<I>> {
         let segment_id = Self::height_to_segment_id(height);
+
```
[the remark this answers is not in the document]
Comment (xdustinface) [reply dash-spv/src/storage/segments.rs:298]: Documented on `get_item` that, by design, it returns `Ok(None)` for truncated slots where `get_items` would return `InvalidArgument`, and that callers must not use `get_item` as a fallback for an error from the range API. Fixed in a7ec7fe5.

File: dash-spv/src/storage/segments.rs [hunk 20]
```diff
@@ -476,6 +569,29 @@ impl<I: Persistable> Segment<I> {
         Ok(())
```
[the remark this answers is not in the document]
Comment (xdustinface) [reply dash-spv/src/storage/segments.rs:587]: Dropped `pub` from `Segment::reset_above` entirely (the method is only called from `SegmentCache::truncate_above` in the same module). Module-private is the most restrictive scope that compiles and matches our "most restrictive visibility possible" rule. Fixed in bf07390f.

File: dash-spv/src/storage/segments.rs [hunk 21]
```diff
@@ -215,6 +223,16 @@ impl<I: Persistable> SegmentCache<I> {
         let end_segment = Self::height_to_segment_id(end - 1);
 
         for segment_id in start_segment..=end_segment {
+            // Same guard as `get_item`: a segment queued for deletion holds
+            // nothing but sentinels by construction. Loading it via
+            // `get_segment_mut` would consume the deletion intent and queue
+            // an all-sentinel rewrite, so refuse the read instead.
```
[the remark this answers is not in the document]
Comment (xdustinface) [reply dash-spv/src/storage/segments.rs:236]: Pushing back on this for now. PR #149 is Phase 1 plumbing and the Phase 3 reorg coordinator that would actually discriminate this error doesn't exist yet, so adding a new public `StorageError` variant speculatively would be a design decision made without a real caller to validate the shape or name (`DataTruncated` vs `NotAvailable` vs something keyed on the queued segment id, etc.). The error message already encodes the case (`get_items range … extends into segment N queued for deletion`), so any callers landing in the Phase 1→3 gap can match on it for diagnostics without us locking in the variant prematurely. Happy to introduce a dedicated variant when Phase 3 lands with a caller that needs to branch on it programmatically. Resolving for now.

**Comment by xdustinface** [comment]
@manki review

File: dash-spv/src/storage/segments.rs [hunk 22]
```diff
@@ -654,6 +788,283 @@ mod tests {
...
+        assert_eq!(cache.tip_height(), Some(ITEMS_PER_SEGMENT + 4));
+
+        cache.truncate_above(ITEMS_PER_SEGMENT - 1).await.unwrap();
+        assert_eq!(cache.tip_height(), Some(ITEMS_PER_SEGMENT - 1));
+
+        let kept = cache.get_items(0..ITEMS_PER_SEGMENT).await.unwrap();
+        assert_eq!(kept.len(), ITEMS_PER_SEGMENT as usize);
+
+        // The dropped segment file should not be on disk after persist.
+        cache.persist(tmp_dir.path()).await;
```
[the remark this answers is not in the document]
Comment (xdustinface) [reply dash-spv/src/storage/segments.rs:867]: Fixed in 6e48d700. The test now persists segment 1 to disk and asserts it exists before calling `truncate_above`, then verifies the file is removed after the truncate+persist cycle. The downstream durability and re-store assertions are unchanged.

File: dash-spv/src/storage/mod.rs (hunk 9 above)
[the remark this answers is not in the document]
Comment (xdustinface) [reply dash-spv/src/storage/mod.rs:322]: Pushing back. This is a design suggestion for Phase 3, not a bug in Phase 1. The reviewer comment itself defers it ("A cleaner design for Phase 3 would...") and notes "cross-storage atomicity is out of scope for Phase 1".

Concretely:

1. The current code is correct. UFCS is the standard, idiomatic Rust mechanism for resolving overlapping trait methods, not a bug.
2. Production callers operate through trait objects (`Arc<dyn BlockHeaderStorage>`, etc.), so they never see the disambiguation issue. Only the test at line 599 hits UFCS, and it does so intentionally to verify the block-header-specific implementation.
3. Adding a fan-out `DiskStorageManager::truncate_above` right now without cross-storage atomicity would create a convenience API that hides a real partial-failure mode. The honest path is to design that as part of Phase 3 (reorg engine), where atomicity semantics can be addressed properly.

Closing as deferred to Phase 3.

**Comment by xdustinface** [comment]
@manki review

**Comment by xdustinface** [comment]
@manki review

File: dash-spv/src/storage/filters.rs [hunk 23]
```diff
@@ -62,4 +78,75 @@ impl FilterStorage for PersistentFilterStorage {
     async fn filter_tip_height(&self) -> StorageResult<u32> {
```
[the remark this answers is not in the document]
Comment (xdustinface) [reply dash-spv/src/storage/filters.rs:78]: Pushing back on this one.

The three higher-level storage types (`PersistentFilterStorage`, `PersistentFilterHeaderStorage`, `PersistentBlockStorage`) delegate directly to `SegmentCache::truncate_above` with no extra logic. The error path is already covered by `test_truncate_below_start_errors` at the segment-cache layer, which is the layer where the validation actually lives.

Adding three near-identical tests at each delegating boundary tests the language (that method calls propagate `Result`) rather than our code. The cited risk (*"a future refactor that adds per-storage guards could silently swallow the error"*) is genuinely hypothetical. If a future change introduces per-storage logic, the test should be added with that change, against the new logic, not pre-emptively against a pass-through.

This is round 7 of polish on a PR the reviewer agrees is correct as written. Closing the thread.

**Comment by xdustinface** [comment]
@manki review

File: dash-spv/src/storage/segments.rs [hunk 24]
```diff
@@ -322,9 +362,80 @@ impl<I: Persistable> SegmentCache<I> {
...
+        // Load the boundary segment first so any disk I/O error is surfaced
+        // before mutating cache state. After this point only infallible
+        // in-memory operations run, so the function cannot leave the cache
+        // in a half-truncated state.
+        if boundary_offset + 1 < Segment::<I>::ITEMS_PER_SEGMENT {
+            let segment = self.get_segment_mut(&boundary_segment_id).await?;
+            segment.reset_above(boundary_offset);
+        }
+
+        for segment_id in (boundary_segment_id + 1)..=max_segment_id {
```
[the remark this answers is not in the document]
Comment (xdustinface) [reply dash-spv/src/storage/segments.rs:411]: Applied: switched to `tokio::fs::remove_file(...).await` in [`505c62f0`](https://github.com/xdustinface/rust-dashcore/pull/149/commits/cdc56382). The atomic-write path was already async, so this restores consistency and avoids stalling executor threads on slow filesystems.

File: dash-spv/src/storage/segments.rs [hunk 25]
```diff
@@ -654,6 +788,287 @@ mod tests {
...
+        let dropped_segment_file = tmp_dir.path().join(FilterHeader::segment_file_name(1));
+        assert!(dropped_segment_file.exists());
+
+        cache.truncate_above(ITEMS_PER_SEGMENT - 1).await.unwrap();
+        assert_eq!(cache.tip_height(), Some(ITEMS_PER_SEGMENT - 1));
+
+        let kept = cache.get_items(0..ITEMS_PER_SEGMENT).await.unwrap();
+        assert_eq!(kept.len(), ITEMS_PER_SEGMENT as usize);
+
+        cache.persist(tmp_dir.path()).await;
```
[the remark this answers is not in the document]
Comment (xdustinface) [reply dash-spv/src/storage/segments.rs:867]: Applied in [`505c62f0`](https://github.com/xdustinface/rust-dashcore/pull/149/commits/505c62f0): swapped the length-only assertion for `assert_eq!(kept, items[0..ITEMS_PER_SEGMENT as usize])` so a zero-fill or stale-segment regression in the boundary-segment branch would actually fail the test.

File: dash-spv/src/storage/segments.rs [hunk 26]
```diff
@@ -654,6 +788,287 @@ mod tests {
         );
```
[the remark this answers is not in the document]
Comment (xdustinface) [reply dash-spv/src/storage/segments.rs:788]: Declining this one. The `target_height == start_height` arithmetic is already exercised by existing tests: `test_truncate_above_within_segment` truncates at height 9 with `start_height = 0` (so `reset_above(9)` zeroes slots 10..N), and `test_truncate_within_segment_persist_reload` covers the same boundary across a persist/reload. The proposed test would only add the offset=0 case, which is a single-slot variation of the same code path. At this point in the review cycle the marginal coverage doesn't justify another commit. Resolving as won't-fix.

**Comment by xdustinface** [comment]
@manki review

**Comment by xdustinface** [comment]
@manki review

**Comment by xdustinface** [comment]
@manki review
